include/libtorrent/packet_pool.hpp

```diff
@@ -0,0 +1,156 @@
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#ifndef TORRENT_PACKET_POOL_HPP
+#define TORRENT_PACKET_POOL_HPP
+
+#include "libtorrent/config.hpp"
+
+#include "libtorrent/aux_/vector.hpp"
+
+namespace libtorrent
+{
+
+	// used for out-of-order incoming packets
+	// as well as sent packets that are waiting to be ACKed
+	struct packet
+	{
+		// the last time this packet was sent
+		time_point send_time;
+
+		// the number of bytes actually allocated in 'buf'
+		std::uint16_t allocated;
+
+		// the size of the buffer 'buf' points to
+		std::uint16_t size;
+
+		// this is the offset to the payload inside the buffer
+		// this is also used as a cursor to describe where the
+		// next payload that hasn't been consumed yet starts
+		std::uint16_t header_size;
+
+		// the number of times this packet has been sent
+		std::uint8_t num_transmissions:6;
+
+		// true if we need to send this packet again. All
+		// outstanding packets are marked as needing to be
+		// resent on timeouts
+		bool need_resend:1;
+
+		// this is set to true for packets that were
+		// sent with the DF bit set (Don't Fragment)
+		bool mtu_probe:1;
+
+	#if TORRENT_USE_ASSERTS
+		int num_fast_resend;
+	#endif
+
+		// the actual packet buffer
+		std::uint8_t buf[1];
+	};
+
+	// single thread packet allocation packet pool
+	// can handle common cases of packet size by 3 pools
+	struct TORRENT_EXTRA_EXPORT packet_pool : single_threaded
+	{
+		packet *acquire(std::size_t allocate)
+		{
+			TORRENT_ASSERT(is_single_thread());
+			TORRENT_ASSERT(allocate <= std::numeric_limits<std::uint16_t>::max());
+
+			packet *p;
+
+			if (allocate <= m_syn_slab.allocate_size) { p = m_syn_slab.try_alloc(); }
+			else if (allocate <= m_mtu_floor_slab.allocate_size) { p = m_mtu_floor_slab.try_alloc(); }
+			else if (allocate <= m_mtu_ceiling_slab.allocate_size) { p = m_mtu_ceiling_slab.try_alloc(); }
+			else { p = static_cast<packet*>(malloc(sizeof(packet) + allocate)); }
+
+			p->allocated = static_cast<std::uint16_t>(allocate);
+			return p;
+		}
+
+		void release(packet *p)
+		{
+			TORRENT_ASSERT(is_single_thread());
+
+			if (p == nullptr) return;
+			std::size_t allocated = p->allocated;
+
+			if (allocated <= m_syn_slab.allocate_size) { m_syn_slab.try_free(p); }
+			else if (allocated <= m_mtu_floor_slab.allocate_size) { m_mtu_floor_slab.try_free(p); }
+			else if (allocated <= m_mtu_ceiling_slab.allocate_size) { m_mtu_ceiling_slab.try_free(p); }
+			else { free(p); }
+		}
+	private:
+
+		template<std::size_t PACKET_LIMIT, std::size_t ALLOCATE_SIZE>
+		struct packet_slab
+		{
+			aux::vector<packet*, std::size_t> m_storage;
+			std::size_t idx{ 0 };
+			static const std::size_t allocate_size{ ALLOCATE_SIZE };
+
+			packet_slab() : m_storage(PACKET_LIMIT, nullptr) { }
+
+			~packet_slab() { flush(); }
+
+			void flush()
+			{
+				while (idx != 0) {
+					free(m_storage[--idx]);
+				}
+			}
+
+			void try_free(packet *p)
+			{
+				if (idx >= PACKET_LIMIT)
```

Comment:
I think this function should do:
```
if (m_storage.size() >= m_limit)
   std::free(p);
else
   m_storage.push_back(p);
```

it's probably a good idea to ``TORRENT_ASSERT`` the packet size is greater than or equal to the slab size as well